src/libfsm/print/dfaops.c

```diff
@@ -0,0 +1,393 @@
+	 * We'll need that in order to be able to use the leaf callback here. */
+	(void) leaf;
+	(void) leaf_opaque;
+
+	/* TODO: we'll need to heed cp for e.g. lx's codegen */
+	(void) cp;
+
+	if (!dfavm_compile_ir(&a, ir, vm_opts)) {
+		return -1;
+	}
+
+	for (op = a.linked; op != NULL; op = op->next) {
+		if (op->num_incoming > 0) {
+			print_label(f, op, opt, dialect);
+		}
+		switch (op->instr) {
+		case VM_OP_STOP:
+			print_cond(f, op, opt, dialect);
+			print_end(f, op, opt,  dialect, op->u.stop.end_bits, ir);
+			break;
+
+		case VM_OP_FETCH:
+			print_fetch(f, opt, dialect);
+			print_end(f, op, opt, dialect, op->u.fetch.end_bits, ir);
+			break;
+
+		case VM_OP_BRANCH:
+			print_cond(f, op, opt, dialect);
+			print_branch(f, op, dialect);
+			break;
+
+		default:
+			assert(!"unreached");
+			break;
+		}
+	}
+
+	dfavm_opasm_finalize_op(&a);
+
+	return 0;
+}
+
+static void
+fsm_print_dfaops_complete(FILE *f, const struct ir *ir, const struct fsm_options *opt, enum dfaops_dialect dialect)
+{
+	/* TODO: currently unused, but must be non-NULL */
+	const char *cp = "";
+
+	assert(f != NULL);
+	assert(ir != NULL);
+	assert(opt != NULL);
+
+	switch (dialect) {
+	case DFAOPS_H:
+	case DFAOPS_MAIN:
+		return;
+	default:
+		break;
+	}
+
+	(void) fsm_print_dfaopsfrag(f, ir, opt, dialect, cp,
+		opt->leaf != NULL ? opt->leaf : leaf, opt->leaf_opaque);
+}
+
+void
+fsm_print_dfaops(FILE *f, const struct fsm *fsm, enum dfaops_dialect dialect)
+{
+	struct ir *ir;
+	const char *prefix;
+	const char *package_prefix;
+
+	assert(f != NULL);
+	assert(fsm != NULL);
+	assert(fsm->opt != NULL);
+
+	ir = make_ir(fsm);
+	if (ir == NULL) {
+		return;
+	}
+
+	/* henceforth, no function should be passed struct fsm *, only the ir and options */
+
+	if (fsm->opt->prefix != NULL) {
+		prefix = fsm->opt->prefix;
+	} else {
+		prefix = "fsm_";
+	}
+
+	if (fsm->opt->fragment) {
+		fsm_print_dfaops_complete(f, ir, fsm->opt, dialect);
+		return;
+	}
+
+	if (fsm->opt->package_prefix != NULL) {
+		package_prefix = fsm->opt->package_prefix;
+	} else {
+		package_prefix = prefix;
+	}
+
+	switch (dialect) {
+
+	case DFAOPS_C:
+		fprintf(f, "#include <stdint.h>\n\n");
+		fprintf(f, "#include \"dfaops.h\"\n\n");
+		fprintf(f, "struct op %sDFA[] = {\n", prefix);
+		break;
+
+	case DFAOPS_H:
+		fprintf(f, "#include <stdint.h>\n\n");
+		fprintf(f, "enum dfaOp { opEOF, opLT, opLE, opEQ, opNE, opGE, opGT, opALWAYS};\n");
+		fprintf(f, "enum actionOp { actionRET, actionGOTO };\n");
+		fprintf(f, "struct op { enum dfaOp op; char c; enum actionOp action; int32_t arg; };\n\n");
+		fprintf(f, "extern struct op %sDFA[];\n", prefix);
+		break;
+
+	case DFAOPS_MAIN:
+		fprintf(f, "#include <stdio.h>\n\n");
+		fprintf(f, "#include <stdlib.h>\n\n");
+		fprintf(f, "#include \"dfaops.h\"\n\n");
+		fprintf(f, "\n");
+
+		switch (fsm->opt->io) {
+		case FSM_IO_PAIR:
+			fprintf(f, "int matchDFA(struct op ops[], char *b, char *e)\n");
```

Comment:
I think these functions should be prefixed with the `prefix`. and probably named just `%smatch`, rather than `%smatchDFA`